### PR TITLE
FEAT/REFAC(plugins): Add classes for process-related functions

### PR DESCRIPTION
**Pull request by davidebeatrici**
For now only `se` (Source Engine) is converted to use the new classes.

### CODE REVIEW

**Comment by davidebeatrici** [comment]
@Krzmbrzl Should the message after the colon start with a capital letter or not?

File: plugins/MumblePluginProcess.hpp [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+
+		iovec out;
+		out.iov_base = dst;
+		out.iov_len  = size;
+
+		const auto ret = process_vm_readv(id, &out, 1, &in, 1, 0);
+
+		return (ret != -1 && static_cast< size_t >(ret) == in.iov_len);
+	}
+#endif
```
Comment (Krzmbrzl) [inline plugins/MumblePluginProcess.hpp:53]: Just make `peek` a pure virtual function and put the platform-specific code in the respective sub-class. That way it'll be easier to find the implementation for a given function (and tidies up the code a bit).

And besides: Implementations shouldn't be in header files (for non-templates)

File: plugins/MumblePluginProcess.hpp [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+				string += ch;
+
+				// Update current time.
+				now = std::chrono::steady_clock::now();
+			}
+		}
+
+		return string;
+	}
```
Comment (Krzmbrzl) [inline plugins/MumblePluginProcess.hpp:137]: Implementations like this should go in a separate cpp file

File: plugins/MumblePluginProcess.hpp [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+		}
+
+		return peekPtr(vTable + (index * pointerSize));
+	}
+
+	inline procptr_t module() { return module(name); }
+#ifdef OS_WINDOWS
+	virtual procptr_t module(const std::string &module) = 0;
+#else
+	procptr_t module(const std::string &module) {
```
Comment (Krzmbrzl) [inline plugins/MumblePluginProcess.hpp:162]: Again: This should be single pure virtual function that is implemented by the respective usb-class

File: plugins/MumblePluginProcess.hpp [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+#	include <sys/uio.h>
+#endif
+
+typedef uint32_t procid_t;
+typedef uint64_t procptr_t;
+
+/// Abstract class.
+/// Only defines stuff that can be used with both Linux and Windows processes.
+class MumblePluginProcess {
```
Comment (Krzmbrzl) [inline plugins/MumblePluginProcess.hpp:28]: I don't like using the term "plugin" when what it really is related to is "positional audio". Especially since there will be actual plugins coming to Mumble in the foreseeable future.

File: plugins/MumblePluginProcess.hpp [hunk 5]
```diff
@@ -0,0 +1,313 @@
+// Copyright 2020 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+#ifndef MUMBLEPLUGINPROCESS_HPP_
```
Comment (Krzmbrzl) [inline plugins/MumblePluginProcess.hpp:6]: Why is this a `.hpp`file while we are using `.h` almost everywhere else for our header files?

File: plugins/se/client.h [hunk 6]
```diff
@@ -42,20 +42,20 @@ struct EntityCacheInfo {
 static int32_t getDataVar(const std::string &name, procptr_t predMap) {
 	while (predMap) {
 		DataMap dataMap;
-		if (!peekProc(predMap, dataMap)) {
+		if (!proc->peek(predMap, dataMap)) {
```
Comment (Krzmbrzl) [inline plugins/se/client.h:45]: where is `proc` defined and initialized?

File: plugins/se/se.cpp [hunk 7]
```diff
@@ -3,8 +3,20 @@
...
+#endif
+#include "../MumblePluginProcessWindows.hpp"
+
+std::unique_ptr< MumblePluginProcess > proc;
+
+#ifdef OS_WINDOWS
+constexpr static bool isWin32 = true;
+#else
+static bool isWin32;
+#endif
```
Comment (Krzmbrzl) [inline plugins/se/se.cpp:19]: 1. Why is one labeled `constexpr` and the other `static`?
2. I would explicitly initialize `isWin32 = false` as well. Relying on default values is bit smelly imo.

**Review by Krzmbrzl — CHANGES_REQUESTED** [review]

**Comment by Krzmbrzl** [comment]
> Should the message after the colon start with a capital letter or not?

For the commit messages? I think I always start them with a capital letter :thinking: 
But in the end I don't think that it matters all that much. I probably won't ask anyone to change their commit message if this is all that is to be criticized :shrug:

File: plugins/MumblePluginProcess.hpp (hunk 1 above)
Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp:53]: The reason why I didn't do it is due to `peek()`'s implementation being the same for both native and Wine processes.

File: plugins/MumblePluginProcess.hpp (hunk 3 above)
Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp:162]: The reason why I didn't do it is due to `module()`'s implementation being the same for both native and Wine processes.

File: plugins/MumblePluginProcess.hpp (hunk 4 above)
Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp:28]: I agree. Does `ProcessUtils` sound good?

File: plugins/se/client.h (hunk 6 above)
Comment (davidebeatrici) [reply plugins/se/client.h:45]: It's defined in `se.cpp`, before the header is included. Definitely not ideal, in future I will probably write proper classes and pass it as `std::shared_ptr`.

File: plugins/se/se.cpp (hunk 7 above)
Comment (davidebeatrici) [reply plugins/se/se.cpp:19]: 1. Both are labeled `static`. The reason why it's `constexpr` on Windows is because the process is guaranteed to be native.
The variable is used in the plugin's headers to determine how to parse some specific parts of the game's assembly.
2. `static` global variables are actually initialized to 0 by default: https://stackoverflow.com/questions/2091499/why-are-global-and-static-variables-initialized-to-their-default-values

File: plugins/MumblePluginProcess.hpp (hunk 5 above)
Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp:6]: Just because it contains the implementation as well.

However, since you suggested to move it to `.cpp` files, I will rename it to `.h`.

One could argue that `.hpp` would be recommended to make it clear that the header is not compatible with C, but as you said we're already using `.h` everywhere.

File: plugins/MumblePluginProcess.hpp (hunk 2 above)
Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp:137]: It's usually done to prevent the functions from being emitted for every source file that makes use of them.

In the case of plugins it shouldn't matter, as they currently all have a single source file. However:

1. There's no guarantee that plugins will not use more than one source file in the future, especially when the interface you're working on will be available.
2. It tidies up the header quite a bit, especially considering the different function signatures (virtual vs non-virtual) for the two operating systems.

**Comment by davidebeatrici** [comment]
I just would like to maintain consistency, unfortunately in `master` there are already commits of mine that don't start with a capital letter...

File: plugins/MumblePluginProcess.hpp (hunk 1 above)
Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp:53]: "native" being "nativ on Linux"? Then I don't see a problem with moving the implementation to the linux class :thinking:

File: plugins/MumblePluginProcess.hpp (hunk 2 above)
Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp:137]: Aren't these 2 arguments for moving the implementation into a separate cpp file? :eyes: 

By doing so you obviously have to make all plugins include that source file but in return you get what I would call a cleaner separation between declaration and impementation (which is what is usually expected when working with C/C++)

File: plugins/MumblePluginProcess.hpp (hunk 4 above)
Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp:28]: Yeah that would be okay. Depending on what exactly the class is meant to be responsible for, `Process` might be an appropriate name in and of itself :thinking:

File: plugins/MumblePluginProcess.hpp (hunk 5 above)
Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp:6]: > Just because it contains the implementation as well.

I think I never saw this kind of reasoning before :thinking: 

I only know that `.hpp` files should be for cpp headers while `.h` files should be for C headers. But I see `.h` being used for cpp as well more often than not and since we're doing it everywhere else already, I'd opt for sticking to this style :)

File: plugins/se/client.h (hunk 6 above)
Comment (Krzmbrzl) [reply plugins/se/client.h:45]: Ah yes - I see :+1:

File: plugins/se/se.cpp (hunk 7 above)
Comment (Krzmbrzl) [reply plugins/se/se.cpp:19]: Oh yes indeed - I always use `static` as the first keyword wherever I am using it. That's why I missed it ^^

I'm not sure if I like the type of a variable to be dependent on the OS this is compiled on. I think in this case I would opt for removing `constexpr` altogether. I do see that there are also benefits to not doing so :shrug: 

Yeah I know that they are initialized to defaults but it makes reading the code slightly more complicated if you have to check whether in this current situation the variable will have a proper default initialization and what that will end up being. Therefore I prefer explicitly initializing such variables in code, so that it is obvious what the value of this variable will be :point_up:

**Comment by Krzmbrzl** [comment]
Yeah consistentcy is of course important but this specific case is pretty minor imo :shrug:

File: plugins/MumblePluginProcess.hpp (hunk 1 above)
Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp:53]: Yes, native Linux processes. The method is the same for processes that are handled by Wine.

I would have to copy the exact same implementation to the Windows class if I moved it to the Linux one.

File: plugins/MumblePluginProcess.hpp (hunk 2 above)
Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp:137]: Yes, indeed.

File: plugins/se/se.cpp (hunk 7 above)
Comment (davidebeatrici) [reply plugins/se/se.cpp:19]: I agree, I will explicitly initialize `isWin32` to `false`.

As for `constexpr`: it tells the compiler that the variable's value is never going to change.
However, I think in this specific case `const` achieves the same behavior.

File: plugins/MumblePluginProcess.hpp (hunk 4 above)
Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp:28]: Anything process-related. Mostly memory reading.

File: plugins/MumblePluginProcess.hpp (hunk 1 above)
Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp:53]: Does wine use a windows instance?

File: plugins/MumblePluginProcess.hpp (hunk 4 above)
Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp:28]: In that case I think I would even prefer `Process`, but `ProcessUtils` is fine as well. Your call :)

File: plugins/se/se.cpp (hunk 7 above)
Comment (Krzmbrzl) [reply plugins/se/se.cpp:19]: I know what it does. And `const` would be just as "bad" as `constexpr` in this case. My problem here is that one time you have a `static bool` and anohter time you have the type `const(expr) static bool` :shrug:

File: plugins/MumblePluginProcess.hpp (hunk 1 above)
Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp:53]: From a memory viewpoint yes, you see the process exactly as it is on Windows.

When the game is running through Wine `MumblePluginProcessWindows` is used.

File: plugins/se/se.cpp (hunk 7 above)
Comment (davidebeatrici) [reply plugins/se/se.cpp:19]: I don't understand what the problem is though. The data type is still `bool`.

File: plugins/MumblePluginProcess.hpp (hunk 1 above)
Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp]: > When the game is running through Wine MumblePluginProcessWindows is used.

Yeah that was what I wanted to know :+1:

Okay in that case, leave the implementation in the base-class (though as I said elsewhere I would move the actual implementation into a separate cpp file) but make it virtual. And the overwrite it in the Windows class. That way you still don't need the preprocessor stuff. That only makes the code harder to parse :)

File: plugins/se/se.cpp (hunk 7 above)
Comment (Krzmbrzl) [reply plugins/se/se.cpp:22]: No it's not. The const-qualifiers and the `static` are also part of the type. So we have the type `static bool` vs `constexpr static bool` for one and the same variable depending on the platform this is compiled on.
Imo this is not a good design as you potentially have to use other preprocessor statements again when working with this variable in order to differentiate the cases.

A variable's type should not change depending on the platform (imo).

File: plugins/se/CMakeLists.txt [hunk 8]
```diff
@@ -3,4 +3,10 @@
 # that can be found in the LICENSE file at the root of the
 # Mumble source tree or at <https://www.mumble.info/LICENSE>.
 
-add_library(se SHARED "se.cpp")
+add_library(se SHARED
+	"se.cpp"
+
+	"../Process.cpp"
+	"../ProcessLinux.cpp"
+	"../ProcessWindows.cpp"
```
Comment (Krzmbrzl) [inline plugins/se/CMakeLists.txt:11]: I think you must not use `ProcessLinux` on Windows platforms. This seems to create an error on our CI

File: plugins/MumblePluginProcess.hpp (hunk 1 above)
Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp]: Any ideas for a name for the file? I was thinking about `ProcessLinuxCommon.cpp` or `ProcessCommonLinux.cpp`.

Since it doesn't define a new class, also `Process_linux.cpp` or `Process_Linux.cpp` could do.

Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp]: As far as I understood we'd end up with 3 files & classes: `Process` (base-class), `WindowsProcess` and `LinuxProcess` :thinking:

Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp]: `Process` (`Process.cpp/.h`), `ProcessWindows` (`ProcessWindows.cpp/.h`) and `ProcessLinux` (`ProcessLinux.cpp/.h`).

You suggested to remove the preprocessor stuff and move the implementation into a separate `.cpp` file, how should that file be called?

Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp]: That should be `Process.cpp` ("separate" in this case means "a different file than the header file")

Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp]: Ah, I see.

Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp]: Unfortunately this cannot work, as you can see from the CI failure.

A preprocessor statement is still required in order to mark the virtual functions as pure for Windows.

Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp]: I ended up moving the WinAPI implementation to `Process`, which I think is the best solution.

Now `ProcessWindows` only refers to the process itself, rather than the OS.

Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp]: I have to say that I am not a big fan of using preprocessor directives everywhere to divide the implementation into chunks. That is just terrible when it comes to readability of the code.
I much prefer using inheritance to solve this issue. If you don't want to have virtual functions in the resulting code, you can apply a small preprocessor trick for choosing the correct type when actually using these classes, but for the implementation this should be separated by having virtual functions which are overriden as needed

Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp]: https://travis-ci.com/github/mumble-voip/mumble/builds/191018159

https://github.com/mumble-voip/mumble/compare/9df6e9d1571844b0e567e416ff699f8b94d54880..1057ca833c39324e4f7339aa8832f32d481c52b1

Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp]: The error your encountered should be easily fixable: https://stackoverflow.com/questions/3065154/undefined-reference-to-vtable

Just define all functions which are not pure virtual and the error should go away :thinking: 

> Right now the only idea I have for removing the preprocessor directives in a relatively clean way is to move the OS-specific functions from Process.cpp to two separate files (one for each OS). Then we can include them at the bottom of Process.cpp.

Well that'd at least be better than having the preprocessor directives all over the place.

> I thought you had the same idea in #4536 (comment).

No. I was really talking about using virtual functions and inheritance. No preprocessor involved (except for finally choosing which one to instantiate).
Basically what you started to implement in the linked diff. That was exactly what I was talking about.

Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp]: Could you write an example, please?

Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp]: You said that the `peekProc` was the same for Windows and Wine processes, right? In that case you'd have:

Process.h:
```
class Process {
public:
    virtual bool peek(const procptr_t address, void *dst, const size_t size);
};
```
Process.cpp:
```
bool Process::peek(const procptr_t address, void *dst, const size_t size) {
    // implementation windows/wine
}
```

ProcessLinux.h
```
class ProcessLinux {
public:
    virtual bool peek(const procptr_t address, void *dst, const size_t size) override;
};
```

ProcessLinux.cpp
```
bool Process::peek(const procptr_t address, void *dst, const size_t size) {
    // implementation Linux
}
```

And then when it actually comes to instantiating one or the other, we could do something like this:
```
#ifdef OS_WINDOWS
    using process_t = ProcessWIndows;
#else 
    using process_t = ProcessLinux
#endif 

    process_t proc(/* whatever arguments it needs */);
```

Or given that even on Linux we might have to use `ProcessWindows` we have to actually use OO polymorphism:
```
    std::unique_ptr<Process> proc;
    if (windows || wine) {
        proc = std::make_unique<ProcessWindows>(/* whatever arguments it needs */);
    } else {
        proc = std::make_unique<ProcessLinux>(/* whatever arguments it needs */);
    };
```

Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp]: `peek()`'s implementation is completely different for Linux (Wine included) and Windows.

Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp]: Well in that case the function's definition would be pure virtual in `Process` and would then be overwritten by both `ProcessWindows` and `ProcessLinux`

Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp]: It can't be done, as I said the two classes are chosen depending on the process rather than the OS.

Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp]: ???
Why exactly can't it be done? It doesn't matter how you choose the classes. You can always encapsulate different implementations by overwriting existing functionality.

Please provide an actual description of what the problems exactly are (an example with one function might be enough) and how you think to solve it

Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp]: Process

On Linux `ProcessLinux` is used when the target process is native. If the process runs through Wine, `ProcessWindows` is used.

On Windows only `ProcessWindows` is used.

Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp]: So the issue ultimately comes down to this:
The `peek` function uses a different implementation on Linux that it does on Windows, regardless of which process is being inspected (wine vs native)?
But on the other hand we do have the `exportedSymbol` function that depends on whether an actual linux process is being inspected vs a windows process (potentially via wine), regardless of the actual operating system it is running on?

Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp]: Exactly!

Comment (Krzmbrzl) [reply plugins/MumblePluginProcess.hpp]: Okay I see.

In that case we simply need another layer of indirection:
We start from our current baseline `Process` class that declares thee `peek` and `exportedSymbol` functions as pure virtual.

Then we have 2 classes that could be called `LinuxHost` and `WindowsHost` that'll implement the `peek` function as required for the respective OS.

And finally we have our well known `ProcessLinux` and `ProcessWIndows` classes that implement the `exportedSymbol` function.

The `*Host` classes inherit from `Process` while the inheritance of the `Process*` classes will depend on the OS that is being compiled on. For that we could do something like this:
```
#ifdef OS_WINDOWS
using host_class = WindowsHost;
#else 
using host_class = LinuxHost;
#endif 

class ProcessLinux : public host_class {
};
```

With this solution we still have some preprocessor statements in our code but in this way I think the code remains much more readable than by doing it in the current way :thinking: 

----

Btw: What about MacOS? Do we even support the overlay on MacOS?

Comment (davidebeatrici) [reply plugins/MumblePluginProcess.hpp]: We support the overlay on macOS: https://github.com/mumble-voip/mumble/blob/master/overlay_gl/init_mac.c

Regarding positional audio plugins, only Link is currently supported: https://github.com/mumble-voip/mumble/blob/master/plugins/link/link-posix.cpp

In theory we can read the memory through `vm_read()` and get the address of an exported symbol by parsing the Mach-O header: https://www.mikeash.com/pyblog/friday-qa-2012-11-30-lets-build-a-mach-o-executable.html

About the classes: your idea is excellent!

**Comment by davidebeatrici** [comment]
If you're wondering why I restored the destructor for `Process`:
> delete called on 'Process' that is abstract but has non-virtual destructor [clang-diagnostic-delete-abstract-non-virtual-dtor]

> delete called on non-final 'ProcessLinux' that has virtual functions but non-virtual destructor [clang-diagnostic-delete-non-abstract-non-virtual-dtor]

> delete called on non-final 'ProcessWindows' that has virtual functions but non-virtual destructor [clang-diagnostic-delete-non-abstract-non-virtual-dtor]

Defining them explicitely for the subclasses is not required. Please let me know whether I should remove them.

File: plugins/se/se.cpp [hunk 9]
```diff
@@ -168,18 +180,31 @@ static int fetch(float *avatarPos, float *avatarFront, float *avatarTop, float *
...
-		return true;
+	for (const auto &name : names) {
+		const auto id = Process::find(name, pids);
+		if (!id) {
+			continue;
+		}
+#ifdef OS_WINDOWS
+		proc.reset(new ProcessWindows(id, name));
+#else
+		isWin32 = Process::isWine(id);
```
Comment (Krzmbrzl) [inline plugins/se/se.cpp:198]:
```suggestion
		isWin32 = HostLinux::isWine(id);
```
Imo this makes it more clear that this is a function that simply is not available on Windows (as in that case `Process` is inheriting from `HostWindows`.

File: plugins/Process.h [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+class Process : public Host {
+protected:
+	std::string m_name;
+	uint8_t m_pointerSize;
+
+public:
+	inline procptr_t module(const std::string &name) const { return Host::module(name); }
+	inline bool peek(const procptr_t address, void *dst, const size_t size) const {
+		return Host::peek(address, dst, size);
+	}
```
[the remark this answers is not in the document]
Comment (davidebeatrici) [reply plugins/Process.h:31]: Unfortunately they aren't, due to the "name hiding" behavior: https://bastian.rieck.me/blog/posts/2016/name_hiding_cxx

File: plugins/se/se.cpp (hunk 9 above)
Comment (davidebeatrici) [reply plugins/se/se.cpp:198]: I agree.

File: plugins/Process.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (Krzmbrzl) [reply plugins/Process.h:31]: Argh... Sometimes I hate cpp :facepalm: 

Nonetheless I think bringing the functions into scope via a `using` statement makes the intent more clear;
```
using Host::peek;
using Host::module;
```

[the remark this answers is not in the document]
Comment (davidebeatrici) [reply plugins/Process.h:31]: I didn't think about it, definitely a much better solution.
